File: packages/frontend/src/types/metrics.ts
export interface WorkflowMetrics {
  workflow_id: string
  workflow_name: string
  total_executions: number
  successful_executions: number
  failed_executions: number
  success_rate: number
  avg_execution_time?: number
  last_execution?: string
  status: 'active' | 'inactive' | 'error'
  time_saved_per_execution_minutes?: number
  time_saved_hours?: number
  is_production?: boolean
  execution_time_trend?: number
  success_rate_trend?: number
}

export interface MetricsTrend {
  execution_trend: number
  success_rate_trend: number
  performance_trend: number
}

export interface ClientMetrics {
  client_id: number
  client_name: string
  total_workflows: number
  active_workflows: number
  total_executions: number
  successful_executions: number
  failed_executions: number
  success_rate: number
  avg_execution_time?: number
  last_activity?: string
  production_executions?: number
  test_executions?: number
  productivity_score?: number
  time_saved_hours?: number
  last_updated?: string  // When metrics were last computed
  trends?: MetricsTrend  // Trend indicators
}

export interface ClientWorkflowMetrics {
  client_id: number
  client_name: string
  workflows: WorkflowMetrics[]
  summary: ClientMetrics
}

export interface AdminMetricsResponse {
  clients: ClientMetrics[]
  total_clients: number
  total_workflows: number
  total_executions: number
  overall_success_rate: number
  total_production_executions?: number
  overall_productivity_score?: number
  total_time_saved_hours?: number  // Overall time saved across all clients
  last_updated?: string  // When admin metrics were last computed
  trends?: MetricsTrend  // Overall system trends
}

export interface MetricsError {
  error: string
  client_id: number
  client_name: string
  details?: string
}

// Historical Metrics Types
export interface MetricsAggregation {
  id: number
  client_id: number
  workflow_id?: number
  period_type: 'DAILY' | 'WEEKLY' | 'MONTHLY'
  period_start: string
  period_end: string
  total_executions: number
  successful_executions: number
  failed_executions: number
  canceled_executions: number
  success_rate: number
  avg_execution_time_seconds?: number
  min_execution_time_seconds?: number
  max_execution_time_seconds?: number
  total_data_size_bytes?: number
  avg_data_size_bytes?: number
  total_workflows?: number
  active_workflows?: number
  most_common_error?: string
  error_count: number
  time_saved_hours?: number
  productivity_score?: number
  computed_at: string
}

export interface HistoricalMetrics {
  client_id: number
  period_type: 'DAILY' | 'WEEKLY' | 'MONTHLY'
  start_date?: string
  end_date?: string
  workflow_id?: number
  aggregations: MetricsAggregation[]
  summary: {
    total_periods: number
    avg_success_rate: number
    total_executions: number
    avg_execution_time: number
    trend_direction: 'UP' | 'DOWN' | 'STABLE'
    trend_percentage: number
  }
}

export interface WorkflowTrendMetrics {
  id: number
  workflow_id: number
  client_id: number
  date: string
  executions_count: number
  success_count: number
  error_count: number
  avg_duration_seconds?: number
  total_duration_seconds?: number
  success_rate_trend?: number
  performance_trend?: number
}

// Enhanced API Response Types
export interface EnhancedClientMetrics extends ClientMetrics {
  recent_trends?: WorkflowTrendMetrics[]
  historical_summary?: {
    last_30_days: MetricsAggregation
    last_7_days: MetricsAggregation
    yesterday: MetricsAggregation
  }
}

export type AggregationPeriod = 'DAILY' | 'WEEKLY' | 'MONTHLY'

export interface MetricsFilters {
  period_type?: AggregationPeriod
  start_date?: string
  end_date?: string
  workflow_id?: number
  include_test_data?: boolean
}

File: packages/backend/app/schemas/metrics.py
"""Metrics schemas"""

from datetime import datetime, date
from typing import List, Optional, Dict, Any
from pydantic import BaseModel


class WorkflowMetrics(BaseModel):
    """Schema for individual workflow metrics"""
    workflow_id: str
    workflow_name: str
    total_executions: int
    successful_executions: int
    failed_executions: int
    success_rate: float
    avg_execution_time: Optional[float] = None
    last_execution: Optional[datetime] = None
    status: str  # 'active', 'inactive', 'error'


class ClientMetrics(BaseModel):
    """Schema for client-level aggregated metrics"""
    client_id: int
    client_name: str
    total_workflows: int
    active_workflows: int
    total_executions: int
    successful_executions: int
    failed_executions: int
    success_rate: float
    avg_execution_time: Optional[float] = None
    last_activity: Optional[datetime] = None
    last_updated: Optional[datetime] = None  # When metrics were last computed


class ClientWorkflowMetrics(BaseModel):
    """Schema for client's workflow-level metrics"""
    client_id: int
    client_name: str
    workflows: List[WorkflowMetrics]
    summary: ClientMetrics


class AdminMetricsResponse(BaseModel):
    """Schema for admin metrics response (all clients)"""
    clients: List[ClientMetrics]
    total_clients: int
    total_workflows: int
    total_executions: int
    overall_success_rate: float
    last_updated: Optional[datetime] = None  # When admin metrics were last computed


class MetricsError(BaseModel):
    """Schema for metrics error response"""
    error: str
    client_id: int
    client_name: str
    details: Optional[str] = None


class MetricsTrend(BaseModel):
    """Schema for metrics trend indicators"""
    execution_trend: float  # Percentage change in executions
    success_rate_trend: float  # Change in success rate
    performance_trend: float  # Performance improvement (negative is slower)


class HistoricalMetrics(BaseModel):
    """Schema for historical metrics data"""
    client_id: int
    workflow_id: Optional[int] = None
    period_type: str  # 'daily', 'weekly', 'monthly'
    start_date: date
    end_date: date
    metrics_data: List[Dict[str, Any]]  # Time series data
    trends: MetricsTrend


class ProductionExecutionFilter(BaseModel):
    """Schema for production execution filtering criteria"""
    exclude_manual: bool = True
    exclude_test_workflows: bool = True
    include_workflow_patterns: Optional[List[str]] = None
    exclude_workflow_patterns: Optional[List[str]] = None